mbed_lstools/windows.py

```diff
@@ -142,65 +147,263 @@ def _determine_valid_non_composite_devices(devices, target_id_usb_id_mount_point
+                if not target_id_usb_id_match:
+                    logger.debug(
+                        "Entry %s has invalid target id pattern %s, skipping",
+                        cached_mount_point_info["mount_point"],
+                        cached_mount_point_info["volume_string"],
+                    )
+                    continue
+
+                target_id_usb_id_mount_point_map[
+                    target_id_usb_id_match.group(1)
+                ] = cached_mount_point_info[
+                    "mount_point"
+                ]
+                disks.pop(index)
+                break
+    return target_id_usb_id_mount_point_map
+
+
+USB_KEY_STRING = "SYSTEM\\CurrentControlSet\\Enum\\USB"
+
+
+def _gen_vid_pid_target_id_usb_id_map(vid_pid_entry_key_string_map, target_id_usb_ids):
+    vid_pid_target_id_usb_id_map = defaultdict(dict)
+    for vid_pid_path, entry_key_strings in vid_pid_entry_key_string_map.items():
+        vid_pid_key_string = "%s\\%s" % (USB_KEY_STRING, vid_pid_path)
+        try:
+            vid_pid_key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, vid_pid_key_string)
+            target_id_usb_id_sub_keys = set([k for k in _iter_keys_as_str(vid_pid_key)])
+        except OSError:
+            logger.debug('VID/PID "%s" not found', vid_pid_key_string)
+            continue
+
+        overlapping_target_id_usb_ids = target_id_usb_id_sub_keys.intersection(
+            set(target_id_usb_ids)
+        )
+        for target_id_usb_id in overlapping_target_id_usb_ids:
+            composite_device_key_string = "%s\\%s" % (
+                vid_pid_key_string, target_id_usb_id
+            )
+            composite_device_key = winreg.OpenKey(vid_pid_key, target_id_usb_id)
+
+            entry_key_string = target_id_usb_id
+            is_prefix = False
+
+            try:
+                new_entry_key_string, _ = winreg.QueryValueEx(
+                    composite_device_key, "ParentIdPrefix"
+                )
+
+                if any(e.startswith(new_entry_key_string) for e in entry_key_strings):
+                    logger.debug(
+                        "Assigning new entry key string of %s to device %s, "
+                        "as found in ParentIdPrefix",
+                        new_entry_key_string,
+                        target_id_usb_id,
+                    )
+                    entry_key_string = new_entry_key_string
+                    is_prefix = True
+            except OSError:
+                logger.debug(
+                    'Device %s did not have a "ParentIdPrefix" key, '
+                    "sticking with %s as entry key string",
+                    composite_device_key_string,
+                    target_id_usb_id,
+                )
+
+            vid_pid_target_id_usb_id_map[vid_pid_path][entry_key_string] = {
+                "target_id_usb_id": target_id_usb_id, "is_prefix": is_prefix
+            }
+    return vid_pid_target_id_usb_id_map
+
+
+def _update_important_bits(
```

Comment:
Let's call this something a bit more descriptive. I'd suggest `_update_device_information` or `_update_device_details`.